File: python/树/783. 二叉搜索树节点最小距离.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def minDiffInBST(self, root: TreeNode) -> int:
        res = []
        def dfs(root):
            if not root: return
            res.append(root.val)
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        res.sort()
        ans = float('+inf')
        for i in range(1,len(res)):
            ans = min(res[i] - res[i-1],ans)
        return ans

# 中序遍历
class Solution(object):
    def minDiffInBST(self, root):
        def dfs(node):
            if node:
                dfs(node.left)
                self.ans = min(self.ans, node.val - self.prev)
                self.prev = node.val
                dfs(node.right)

        self.prev = float('-inf')
        self.ans = float('inf')
        dfs(root)
        return self.ans

File: go/tree/783. 二叉搜索树节点最小距离.go
package tree

import "math"

func minDiffInBST(root *TreeNode) int {
	pre, cur := math.MinInt32, math.MaxInt32
	var dfs func(root *TreeNode)
	dfs = func(root *TreeNode) {
		if root == nil {
			return
		}
		dfs(root.Left)
		cur = min(cur, root.Val-pre)
		pre = root.Val
		dfs(root.Right)
	}
	dfs(root)
	return cur
}

func min(A, B int) int {
	if A < B {
		return A
	} else {
		return B
	}
}
